File: seed_learnings.py
import json
import re
import urllib.request
from html import unescape
from datetime import datetime


ARCHIVE_URL = "https://thedailybrief.zerodha.com/api/v1/archive?sort=new&limit=8"


def fetch(url):
    with urllib.request.urlopen(url, timeout=30) as response:
        return response.read().decode("utf-8", errors="ignore")


def strip_tags(html):
    text = re.sub(r"<script.*?>.*?</script>", "", html, flags=re.S | re.I)
    text = re.sub(r"<style.*?>.*?</style>", "", text, flags=re.S | re.I)
    text = re.sub(r"<[^>]+>", " ", text)
    text = unescape(text)
    text = re.sub(r"\s+", " ", text)
    return text.strip()


def extract_between(html, marker):
    if marker not in html:
        return ""
    start = html.find(marker)
    snippet = html[start:start + 200000]
    return snippet


def extract_article_text(html):
    snippet = extract_between(html, 'available-content')
    if not snippet:
        snippet = extract_between(html, 'body markup')
    return strip_tags(snippet)


def extract_title(html):
    match = re.search(r'<h1 class="post-title"[^>]*>(.*?)</h1>', html, flags=re.S | re.I)
    if match:
        return strip_tags(match.group(1))
    match = re.search(r"<title>(.*?)</title>", html, flags=re.S | re.I)
    return strip_tags(match.group(1)) if match else "Today I Learned"


def extract_date(html):
    match = re.search(r"<time[^>]*datetime=\"([^\"]+)\"", html, flags=re.S | re.I)
    if match:
        return match.group(1)
    return datetime.utcnow().isoformat()


def build_learning(text):
    paragraphs = [p.strip() for p in text.split(". ") if len(p.strip()) >= 60]
    if paragraphs:
        return ". ".join(paragraphs[:4]) + "."
    return text[:600]


def main():
    with open("learnings.json", "r", encoding="utf-8") as f:
        existing = json.load(f)
    if isinstance(existing, list) and len(existing) >= 5:
        print("Seed not needed.")
        return

    archive = json.loads(fetch(ARCHIVE_URL))
    learnings = []

    for article in archive:
        url = article.get("canonical_url")
        if not url:
            continue
        try:
            html = fetch(url)
            content = extract_article_text(html)
            if not content:
                continue
            learning = build_learning(content)
            learnings.append({
                "learning": learning,
                "articleUrl": url,
                "title": extract_title(html),
                "date": extract_date(html)
            })
        except Exception:
            continue

    if learnings:
        with open("learnings.json", "w", encoding="utf-8") as f:
            json.dump(learnings, f, indent=2)
        print(f"Seeded {len(learnings)} learnings.")
    else:
        print("No learnings seeded.")


if __name__ == "__main__":
    main()

File: seed-learnings.js
const axios = require('axios');
const cheerio = require('cheerio');
const fs = require('fs');

function extractParagraphs(text) {
  return (text || '')
    .split('\n')
    .map((line) => line.replace(/\s+/g, ' ').trim())
    .filter((line) => line.length >= 60);
}

function buildFreeLearning(articleText) {
  const paragraphs = extractParagraphs(articleText);
  if (paragraphs.length >= 2) {
    return paragraphs.slice(0, 3).join('\n\n');
  }

  const sentences = (articleText || '')
    .replace(/\s+/g, ' ')
    .split('. ')
    .map((s) => s.trim())
    .filter((s) => s.length >= 40);

  return sentences.slice(0, 4).join('. ') + (sentences.length ? '.' : '');
}

async function fetchArticle(url) {
  const response = await axios.get(url);
  const $ = cheerio.load(response.data);
  const title = $('h1.post-title').text().trim() || $('title').text().trim();
  const date = $('time').attr('datetime') || $('time').text().trim();
  const articleText = $('.available-content').text().trim() || $('.body.markup').text().trim();
  return { title, date, articleText };
}

async function seedLearnings() {
  const existing = JSON.parse(fs.readFileSync('learnings.json', 'utf-8'));
  if (Array.isArray(existing) && existing.length >= 5) {
    console.log('Seed not needed. Learnings already populated.');
    return;
  }

  const response = await axios.get('https://thedailybrief.zerodha.com/api/v1/archive?sort=new&limit=8');
  const articles = response.data || [];
  const seeded = [];

  for (const article of articles) {
    try {
      const { title, date, articleText } = await fetchArticle(article.canonical_url);
      if (!articleText) continue;

      const learning = buildFreeLearning(articleText);
      if (!learning) continue;

      seeded.push({
        learning,
        articleUrl: article.canonical_url,
        title: title || article.title || 'Today I Learned',
        date: date || article.post_date || new Date().toISOString()
      });
    } catch (error) {
      console.error(`Seed failed for ${article.canonical_url}:`, error.message);
    }
  }

  if (!seeded.length) {
    console.log('No learnings seeded.');
    return;
  }

  fs.writeFileSync('learnings.json', JSON.stringify(seeded, null, 2));
  console.log(`Seeded ${seeded.length} learnings.`);
}

seedLearnings().catch((err) => {
  console.error('Seed error:', err);
  process.exit(1);
});
